kratos/tests/cpp_tests/utilities/test_geometry_utils.cpp

```diff
@@ -328,5 +329,240 @@ namespace Testing {
+
+        array_1d<double, 3> pressure_gradient;
+        GeometryUtils::EvaluateHistoricalVariableGradientAtGaussPoint(pressure_gradient, tetrahedra, PRESSURE, DN_DX);
+        KRATOS_CHECK_VECTOR_NEAR(check_pressure_gradient, pressure_gradient, 1e-15);
+
+        // testing for previous step
+        tetrahedra[0].FastGetSolutionStepValue(PRESSURE, 1) = 10.0;
+        tetrahedra[1].FastGetSolutionStepValue(PRESSURE, 1) = 20.0;
+        tetrahedra[2].FastGetSolutionStepValue(PRESSURE, 1) = 30.0;
+        tetrahedra[3].FastGetSolutionStepValue(PRESSURE, 1) = 40.0;
+
+        array_1d<double, 3> check_old_pressure_gradient;
+        check_old_pressure_gradient[0] = DN_DX(0, 0) * 10.0 + DN_DX(1, 0) * 20.0 + DN_DX(2, 0) * 30.0 + DN_DX(3, 0) * 40.0;
+        check_old_pressure_gradient[1] = DN_DX(0, 1) * 10.0 + DN_DX(1, 1) * 20.0 + DN_DX(2, 1) * 30.0 + DN_DX(3, 1) * 40.0;
+        check_old_pressure_gradient[2] = DN_DX(0, 2) * 10.0 + DN_DX(1, 2) * 20.0 + DN_DX(2, 2) * 30.0 + DN_DX(3, 2) * 40.0;
+
+        array_1d<double, 3> old_pressure_gradient;
+        GeometryUtils::EvaluateHistoricalVariableGradientAtGaussPoint(old_pressure_gradient, tetrahedra, PRESSURE, DN_DX, 1);
+        KRATOS_CHECK_VECTOR_NEAR(check_old_pressure_gradient, old_pressure_gradient, 1e-15);
+    }
+
+    KRATOS_TEST_CASE_IN_SUITE(EvaluateHistoricalVariableGradientAtGaussPointArray, KratosCoreFastSuite)
+    {
+        VariablesList::Pointer current_variable_list = Kratos::make_intrusive<VariablesList>();
+        current_variable_list->Add(VELOCITY);
+
+        Tetrahedra3D4 <NodeType> tetrahedra = GenerateExampleTetrahedra();
+
+        tetrahedra[0].SetSolutionStepVariablesList(current_variable_list);
+        tetrahedra[1].SetSolutionStepVariablesList(current_variable_list);
+        tetrahedra[2].SetSolutionStepVariablesList(current_variable_list);
+        tetrahedra[3].SetSolutionStepVariablesList(current_variable_list);
+
+        tetrahedra[0].SetBufferSize(2);
+        tetrahedra[1].SetBufferSize(2);
+        tetrahedra[2].SetBufferSize(2);
+        tetrahedra[3].SetBufferSize(2);
+
+        // Computing the info
+        BoundedMatrix<double,4,3> DN_DX;
+        array_1d<double,4> N;
+        double volume;
+
+        GeometryUtils::CalculateGeometryData(tetrahedra, DN_DX, N, volume);
+
+        // testing for current step
+        array_1d<double, 3> n0(3, 1.0);
+        array_1d<double, 3> n1(3, 2.0);
+        array_1d<double, 3> n2(3, 3.0);
+        array_1d<double, 3> n3(3, 4.0);
+        tetrahedra[0].FastGetSolutionStepValue(VELOCITY) = n0;
+        tetrahedra[1].FastGetSolutionStepValue(VELOCITY) = n1;
+        tetrahedra[2].FastGetSolutionStepValue(VELOCITY) = n2;
+        tetrahedra[3].FastGetSolutionStepValue(VELOCITY) = n3;
+
+        BoundedMatrix<double, 3, 3> check_velocity_gradient;
+        check_velocity_gradient(0, 0) = n0[0] * DN_DX(0, 0) + n1[0] * DN_DX(1, 0) + n2[0] * DN_DX(2, 0) + n3[0] * DN_DX(3, 0);
```

Comment:
i would argue that this check will be much shorter and mantainable if you stored the ns in a matrix, then you could compute the check using the prod function. (i'll accept it as it is, take it simply as a constructive comment for the future)

Reply:
thanks :)